adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m

```diff
@@ -13,36 +13,78 @@
 // limitations under the License.
 
 #import "GADMediationAdapterPangle.h"
-
 #import "GADMediationAdapterPangleConstants.h"
+#import <BUAdSDK/BUAdSDK.h>
+#import "GADPangleRTBBannerRenderer.h"
+#import "GADPangleRTBInterstitialRenderer.h"
+#import "GADPangleRTBRewardedlRenderer.h"
+#import "GADPangleRTBNativeRenderer.h"
+#import "GADMediationAdapterPangleConstants.h"
+#import "GADMAdapterPangleUtils.h"
+
+@interface GADMediationAdapterPangle ()
+
+@property (nonatomic, strong) GADPangleRTBBannerRenderer *bannerRenderer;
+@property (nonatomic, strong) GADPangleRTBInterstitialRenderer *interstitialRenderer;
+@property (nonatomic, strong) GADPangleRTBRewardedlRenderer *rewardedlRenderer;
+@property (nonatomic, strong) GADPangleRTBNativeRenderer *navtiveRenderer;
+
+@end
 
 @implementation GADMediationAdapterPangle
 
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:(nonnull GADRTBSignalCompletionHandler)completionHandler {
-  // TODO: Start signal generation and call completionHandler when the signal generation finishes or fails.
-  completionHandler("", nil);
+    NSString *signals = [BUAdSDKManager mopubBiddingToken];
+    if (completionHandler) {
+        completionHandler(signals,nil);
+    }
 }
 
 + (void)setUpWithConfiguration:(nonnull GADMediationServerConfiguration *)configuration
              completionHandler:(nonnull GADMediationAdapterSetUpCompletionBlock)completionHandler {
-  // TODO: Set up Pangle ad network SDK and perform any necessary prefetching or configuration work.
-  completionHandler(nil);
+    NSMutableSet *appIds = [NSMutableSet new];
+    for (GADMediationCredentials *cred in configuration.credentials) {
+        NSString *appId = cred.settings[GADMAdapterPangleAppID];
+        if (appId && [appId isKindOfClass:[NSString class]] && appId.length) {
+            [appIds addObject:appId];
+        }
+    }
+    
+    if (appIds.count < 1) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorMissingServerParameters, @"pangle mediation configuration did not contain a valid app id");
```

Comment:
Nit: "Pangle mediation configurations did not contain a valid App ID."

Reply:
solved